Add function to extract a sub model object from a full model

isorrentino opened the pull request:
The PR implements the possibility to create a new `iDynTree::Model` containing links and joints composing the subModelTraversalstarting from a full model and a sub traversal of the full model. Therefore, the new model will contain joints and links composing the subModelTraversal, in the same order, and all the additional frames originally attached to considered links.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/robotology/idyntree?pullRequest=1011) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/robotology/idyntree?pullRequest=1011) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/robotology/idyntree?pullRequest=1011) it.</sub>

traversaro commented on src/model/src/ModelTransformers.cpp:
```diff
@@ -555,4 +555,73 @@ bool createModelWithNormalizedJointNumbering(const Model& model,
+bool extractSubModel(const iDynTree::Model& fullModel, const iDynTree::Traversal& subModelTraversal,
+                     iDynTree::Model& outputSubModel)
+{
+    if (&subModelTraversal == NULL)
+    {
+        std::cerr << "[ERROR] subModelTraversal is null" << std::endl;
+        return false;
+    }
```
We tipically do not check the validity of references as to pass an invalid reference you would need to dereference an invalid pointer, and that would raise an error before (I think).

isorrentino commented:
I also added the bindings generation for `SubModel.h` and `ModelTransformers.h`.

The PR is ready for review.

cc @traversaro

traversaro commented on src/high-level/tests/KinDynComputationsUnitTest.cpp:
```diff
@@ -681,6 +833,14 @@ int main()
+    testModelConsistencyAllRepresentations("bigman.urdf");
+    testModelConsistencyAllRepresentations("icub_skin_frames.urdf");
+    testModelConsistencyAllRepresentations("iCubGenova02.urdf");
+    testModelConsistencyAllRepresentations("icalibrate.urdf");
```
It is intended to only have `testSubModelConsistencyAllRepresentations` for one of the models? If so, probably we can remove the `testModelConsistencyAllRepresentations` for the other models.

isorrentino replied:
Maybe one model is enough, I can remove the others.

traversaro commented:
Thanks! Can I merge with squash?

isorrentino commented:
I changed the model I was passing to `testSubModelConsistencyAllRepresentations`, from `threeLinks` to `iCubGenova02` and the test is giving a segmentation fault. I printed the full model and the sub model and I discovered that it fails here:

https://github.com/robotology/idyntree/blob/aab291d422ef3c314324722149e6e697827f83ce/src/high-level/tests/KinDynComputationsUnitTest.cpp#L636-L640

The two models are:

<details>
<summary style="font-size:14px">reduced model</summary>

```
Reduced model
  Links: 
    [0] root_link
    [1] r_hip_1
    [2] l_hip_1
    [3] torso_1
    [4] torso_2
    [5] chest
    [6] neck_1
    [7] l_shoulder_1
    [8] r_shoulder_1
    [9] r_shoulder_2
    [10] r_shoulder_3
    [11] l_shoulder_2
    [12] l_shoulder_3
    [13] l_upper_arm
    [14] l_elbow_1
    [15] neck_2
    [16] head
    [17] l_hip_2
    [18] l_hip_3
    [19] l_upper_leg
    [20] l_lower_leg
    [21] l_ankle_1
    [22] l_ankle_2
    [23] l_foot
    [24] r_hip_2
    [25] r_hip_3
    [26] r_upper_leg
    [27] r_lower_leg
    [28] r_ankle_1
    [29] r_ankle_2
    [30] r_foot
  Frames: 
    [31] base_link --> root_link
    [32] root_link_ems_acc_eb5 --> root_link
    [33] root_link_ems_gyro_eb5 --> root_link
    [34] chest_ems_acc_eb1 --> chest
    [35] chest_ems_gyro_eb1 --> chest
    [36] chest_ems_acc_eb2 --> chest
    [37] chest_ems_gyro_eb2 --> chest
    [38] chest_ems_acc_eb3 --> chest
    [39] chest_ems_gyro_eb3 --> chest
    [40] chest_ems_acc_eb4 --> chest
    [41] chest_ems_gyro_eb4 --> chest
    [42] chest_mtb_acc_0b7 --> chest
    [43] chest_mtb_acc_0b8 --> chest
    [44] chest_mtb_acc_0b9 --> chest
    [45] chest_mtb_acc_0b10 --> chest
    [46] l_upper_arm_mtb_acc_1b10 --> l_upper_arm
    [47] l_upper_arm_mtb_acc_1b11 --> l_upper_arm
    [48] l_upper_arm_mtb_acc_1b12 --> l_upper_arm
    [49] l_upper_arm_mtb_acc_1b13 --> l_upper_arm
    [50] l_arm_ft_sensor --> l_upper_arm
    [51] l_forearm_mtb_acc_1b7 --> l_elbow_1
    [52] l_forearm_mtb_acc_1b8 --> l_elbow_1
    [53] l_forearm_mtb_acc_1b9 --> l_elbow_1
    [54] imu_frame --> head
    [55] l_leg_ft_sensor --> l_hip_3
    [56] l_upper_leg_ems_acc_eb6 --> l_upper_leg
    [57] l_upper_leg_ems_acc_eb10 --> l_upper_leg
    [58] l_upper_leg_ems_gyro_eb6 --> l_upper_leg
    [59] l_upper_leg_ems_gyro_eb10 --> l_upper_leg
    [60] l_upper_leg_mtb_acc_10b1 --> l_upper_leg
    [61] l_upper_leg_mtb_acc_10b2 --> l_upper_leg
    [62] l_upper_leg_mtb_acc_10b3 --> l_upper_leg
    [63] l_upper_leg_mtb_acc_10b4 --> l_upper_leg
    [64] l_upper_leg_mtb_acc_10b5 --> l_upper_leg
    [65] l_upper_leg_mtb_acc_10b6 --> l_upper_leg
    [66] l_upper_leg_mtb_acc_10b7 --> l_upper_leg
    [67] l_lower_leg_ems_acc_eb7 --> l_lower_leg
    [68] l_lower_leg_ems_gyro_eb7 --> l_lower_leg
    [69] l_lower_leg_mtb_acc_10b8 --> l_lower_leg
    [70] l_lower_leg_mtb_acc_10b9 --> l_lower_leg
    [71] l_lower_leg_mtb_acc_10b10 --> l_lower_leg
    [72] l_lower_leg_mtb_acc_10b11 --> l_lower_leg
    [73] l_foot_mtb_acc_10b12 --> l_foot
    [74] l_foot_mtb_acc_10b13 --> l_foot
    [75] l_sole --> l_foot
    [76] l_foot_dh_frame --> l_foot
    [77] l_foot_ft_sensor --> l_foot
    [78] r_leg_ft_sensor --> r_hip_3
    [79] r_upper_leg_ems_acc_eb8 --> r_upper_leg
    [80] r_upper_leg_ems_acc_eb11 --> r_upper_leg
    [81] r_upper_leg_ems_gyro_eb8 --> r_upper_leg
    [82] r_upper_leg_ems_gyro_eb11 --> r_upper_leg
    [83] r_upper_leg_mtb_acc_11b1 --> r_upper_leg
    [84] r_upper_leg_mtb_acc_11b2 --> r_upper_leg
    [85] r_upper_leg_mtb_acc_11b3 --> r_upper_leg
    [86] r_upper_leg_mtb_acc_11b4 --> r_upper_leg
    [87] r_upper_leg_mtb_acc_11b5 --> r_upper_leg
    [88] r_upper_leg_mtb_acc_11b6 --> r_upper_leg
    [89] r_upper_leg_mtb_acc_11b7 --> r_upper_leg
    [90] r_lower_leg_ems_acc_eb9 --> r_lower_leg
    [91] r_lower_leg_ems_gyro_eb9 --> r_lower_leg
    [92] r_lower_leg_mtb_acc_11b8 --> r_lower_leg
    [93] r_lower_leg_mtb_acc_11b9 --> r_lower_leg
    [94] r_lower_leg_mtb_acc_11b10 --> r_lower_leg
    [95] r_lower_leg_mtb_acc_11b11 --> r_lower_leg
    [96] r_foot_mtb_acc_11b12 --> r_foot
    [97] r_foot_mtb_acc_11b13 --> r_foot
    [98] r_sole --> r_foot
    [99] r_foot_dh_frame --> r_foot
    [100] r_foot_ft_sensor --> r_foot
  Joints: 
    [0] r_hip_pitch (dofs: 1) : root_link<-->r_hip_1
    [1] l_hip_pitch (dofs: 1) : root_link<-->l_hip_1
    [2] torso_pitch (dofs: 1) : root_link<-->torso_1
    [3] torso_roll (dofs: 1) : torso_1<-->torso_2
    [4] torso_yaw (dofs: 1) : torso_2<-->chest
    [5] neck_pitch (dofs: 1) : chest<-->neck_1
    [6] l_shoulder_pitch (dofs: 1) : chest<-->l_shoulder_1
    [7] r_shoulder_pitch (dofs: 1) : chest<-->r_shoulder_1
    [8] r_shoulder_roll (dofs: 1) : r_shoulder_1<-->r_shoulder_2
    [9] r_shoulder_yaw (dofs: 1) : r_shoulder_2<-->r_shoulder_3
    [10] l_shoulder_roll (dofs: 1) : l_shoulder_1<-->l_shoulder_2
    [11] l_shoulder_yaw (dofs: 1) : l_shoulder_2<-->l_shoulder_3
    [12] l_arm_ft_sensor (dofs: 0) : l_shoulder_3<-->l_upper_arm
    [13] l_elbow (dofs: 1) : l_upper_arm<-->l_elbow_1
    [14] neck_roll (dofs: 1) : neck_1<-->neck_2
    [15] neck_yaw (dofs: 1) : neck_2<-->head
    [16] l_hip_roll (dofs: 1) : l_hip_1<-->l_hip_2
    [17] l_leg_ft_sensor (dofs: 0) : l_hip_2<-->l_hip_3
    [18] l_hip_yaw (dofs: 1) : l_hip_3<-->l_upper_leg
    [19] l_knee (dofs: 1) : l_upper_leg<-->l_lower_leg
    [20] l_ankle_pitch (dofs: 1) : l_lower_leg<-->l_ankle_1
    [21] l_ankle_roll (dofs: 1) : l_ankle_1<-->l_ankle_2
    [22] l_foot_ft_sensor (dofs: 0) : l_ankle_2<-->l_foot
    [23] r_hip_roll (dofs: 1) : r_hip_1<-->r_hip_2
    [24] r_leg_ft_sensor (dofs: 0) : r_hip_2<-->r_hip_3
    [25] r_hip_yaw (dofs: 1) : r_hip_3<-->r_upper_leg
    [26] r_knee (dofs: 1) : r_upper_leg<-->r_lower_leg
    [27] r_ankle_pitch (dofs: 1) : r_lower_leg<-->r_ankle_1
    [28] r_ankle_roll (dofs: 1) : r_ankle_1<-->r_ankle_2
    [29] r_foot_ft_sensor (dofs: 0) : r_ankle_2<-->r_foot
```
</details>

<details>
<summary style="font-size:14px">reduced model</summary>
```
  Links: 
    [0] root_link
    [1] r_hip_1
    [2] l_hip_1
    [3] torso_1
    [4] torso_2
    [5] chest
    [6] neck_1
    [7] l_shoulder_1
    [8] r_shoulder_1
    [9] r_shoulder_2
    [10] r_shoulder_3
    [11] r_upper_arm
    [12] r_elbow_1
    [13] l_shoulder_2
    [14] l_shoulder_3
    [15] l_upper_arm
    [16] l_elbow_1
    [17] neck_2
    [18] head
    [19] l_hip_2
    [20] l_hip_3
    [21] l_upper_leg
    [22] l_lower_leg
    [23] l_ankle_1
    [24] l_ankle_2
    [25] l_foot
    [26] r_hip_2
    [27] r_hip_3
    [28] r_upper_leg
    [29] r_lower_leg
    [30] r_ankle_1
    [31] r_ankle_2
    [32] r_foot
  Frames: 
    [33] base_link --> root_link
    [34] root_link_ems_acc_eb5 --> root_link
    [35] root_link_ems_gyro_eb5 --> root_link
    [36] chest_ems_acc_eb1 --> chest
    [37] chest_ems_gyro_eb1 --> chest
    [38] chest_ems_acc_eb2 --> chest
    [39] chest_ems_gyro_eb2 --> chest
    [40] chest_ems_acc_eb3 --> chest
    [41] chest_ems_gyro_eb3 --> chest
    [42] chest_ems_acc_eb4 --> chest
    [43] chest_ems_gyro_eb4 --> chest
    [44] chest_mtb_acc_0b7 --> chest
    [45] chest_mtb_acc_0b8 --> chest
    [46] chest_mtb_acc_0b9 --> chest
    [47] chest_mtb_acc_0b10 --> chest
    [48] r_upper_arm_mtb_acc_2b10 --> r_upper_arm
    [49] r_upper_arm_mtb_acc_2b11 --> r_upper_arm
    [50] r_upper_arm_mtb_acc_2b12 --> r_upper_arm
    [51] r_upper_arm_mtb_acc_2b13 --> r_upper_arm
    [52] r_forearm_mtb_acc_2b7 --> r_elbow_1
    [53] r_forearm_mtb_acc_2b8 --> r_elbow_1
    [54] r_forearm_mtb_acc_2b9 --> r_elbow_1
    [55] l_upper_arm_mtb_acc_1b10 --> l_upper_arm
    [56] l_upper_arm_mtb_acc_1b11 --> l_upper_arm
    [57] l_upper_arm_mtb_acc_1b12 --> l_upper_arm
    [58] l_upper_arm_mtb_acc_1b13 --> l_upper_arm
    [59] l_forearm_mtb_acc_1b7 --> l_elbow_1
    [60] l_forearm_mtb_acc_1b8 --> l_elbow_1
    [61] l_forearm_mtb_acc_1b9 --> l_elbow_1
    [62] imu_frame --> head
    [63] l_upper_leg_ems_acc_eb6 --> l_upper_leg
    [64] l_upper_leg_ems_acc_eb10 --> l_upper_leg
    [65] l_upper_leg_ems_gyro_eb6 --> l_upper_leg
    [66] l_upper_leg_ems_gyro_eb10 --> l_upper_leg
    [67] l_upper_leg_mtb_acc_10b1 --> l_upper_leg
    [68] l_upper_leg_mtb_acc_10b2 --> l_upper_leg
    [69] l_upper_leg_mtb_acc_10b3 --> l_upper_leg
    [70] l_upper_leg_mtb_acc_10b4 --> l_upper_leg
    [71] l_upper_leg_mtb_acc_10b5 --> l_upper_leg
    [72] l_upper_leg_mtb_acc_10b6 --> l_upper_leg
    [73] l_upper_leg_mtb_acc_10b7 --> l_upper_leg
    [74] l_lower_leg_ems_acc_eb7 --> l_lower_leg
    [75] l_lower_leg_ems_gyro_eb7 --> l_lower_leg
    [76] l_lower_leg_mtb_acc_10b8 --> l_lower_leg
    [77] l_lower_leg_mtb_acc_10b9 --> l_lower_leg
    [78] l_lower_leg_mtb_acc_10b10 --> l_lower_leg
    [79] l_lower_leg_mtb_acc_10b11 --> l_lower_leg
    [80] l_foot_mtb_acc_10b12 --> l_foot
    [81] l_foot_mtb_acc_10b13 --> l_foot
    [82] l_sole --> l_foot
    [83] l_foot_dh_frame --> l_foot
    [84] r_upper_leg_ems_acc_eb8 --> r_upper_leg
    [85] r_upper_leg_ems_acc_eb11 --> r_upper_leg
    [86] r_upper_leg_ems_gyro_eb8 --> r_upper_leg
    [87] r_upper_leg_ems_gyro_eb11 --> r_upper_leg
    [88] r_upper_leg_mtb_acc_11b1 --> r_upper_leg
    [89] r_upper_leg_mtb_acc_11b2 --> r_upper_leg
    [90] r_upper_leg_mtb_acc_11b3 --> r_upper_leg
    [91] r_upper_leg_mtb_acc_11b4 --> r_upper_leg
    [92] r_upper_leg_mtb_acc_11b5 --> r_upper_leg
    [93] r_upper_leg_mtb_acc_11b6 --> r_upper_leg
    [94] r_upper_leg_mtb_acc_11b7 --> r_upper_leg
    [95] r_lower_leg_ems_acc_eb9 --> r_lower_leg
    [96] r_lower_leg_ems_gyro_eb9 --> r_lower_leg
    [97] r_lower_leg_mtb_acc_11b8 --> r_lower_leg
    [98] r_lower_leg_mtb_acc_11b9 --> r_lower_leg
    [99] r_lower_leg_mtb_acc_11b10 --> r_lower_leg
    [100] r_lower_leg_mtb_acc_11b11 --> r_lower_leg
    [101] r_foot_mtb_acc_11b12 --> r_foot
    [102] r_foot_mtb_acc_11b13 --> r_foot
    [103] r_sole --> r_foot
    [104] r_foot_dh_frame --> r_foot
    [105] l_arm_ft_sensor --> l_upper_arm
    [106] r_arm_ft_sensor --> r_upper_arm
    [107] l_leg_ft_sensor --> l_hip_3
    [108] r_leg_ft_sensor --> r_hip_3
    [109] l_foot_ft_sensor --> l_foot
    [110] r_foot_ft_sensor --> r_foot
  Joints: 
    [0] r_hip_pitch (dofs: 1) : root_link<-->r_hip_1
    [1] l_hip_pitch (dofs: 1) : root_link<-->l_hip_1
    [2] torso_pitch (dofs: 1) : root_link<-->torso_1
    [3] torso_roll (dofs: 1) : torso_1<-->torso_2
    [4] torso_yaw (dofs: 1) : torso_2<-->chest
    [5] neck_pitch (dofs: 1) : chest<-->neck_1
    [6] l_shoulder_pitch (dofs: 1) : chest<-->l_shoulder_1
    [7] r_shoulder_pitch (dofs: 1) : chest<-->r_shoulder_1
    [8] r_shoulder_roll (dofs: 1) : r_shoulder_1<-->r_shoulder_2
    [9] r_shoulder_yaw (dofs: 1) : r_shoulder_2<-->r_shoulder_3
    [10] r_elbow (dofs: 1) : r_upper_arm<-->r_elbow_1
    [11] l_shoulder_roll (dofs: 1) : l_shoulder_1<-->l_shoulder_2
    [12] l_shoulder_yaw (dofs: 1) : l_shoulder_2<-->l_shoulder_3
    [13] l_elbow (dofs: 1) : l_upper_arm<-->l_elbow_1
    [14] neck_roll (dofs: 1) : neck_1<-->neck_2
    [15] neck_yaw (dofs: 1) : neck_2<-->head
    [16] l_hip_roll (dofs: 1) : l_hip_1<-->l_hip_2
    [17] l_hip_yaw (dofs: 1) : l_hip_3<-->l_upper_leg
    [18] l_knee (dofs: 1) : l_upper_leg<-->l_lower_leg
    [19] l_ankle_pitch (dofs: 1) : l_lower_leg<-->l_ankle_1
    [20] l_ankle_roll (dofs: 1) : l_ankle_1<-->l_ankle_2
    [21] r_hip_roll (dofs: 1) : r_hip_1<-->r_hip_2
    [22] r_hip_yaw (dofs: 1) : r_hip_3<-->r_upper_leg
    [23] r_knee (dofs: 1) : r_upper_leg<-->r_lower_leg
    [24] r_ankle_pitch (dofs: 1) : r_lower_leg<-->r_ankle_1
    [25] r_ankle_roll (dofs: 1) : r_ankle_1<-->r_ankle_2
    [26] r_arm_ft_sensor (dofs: 0) : r_shoulder_3<-->r_upper_arm
    [27] l_arm_ft_sensor (dofs: 0) : l_shoulder_3<-->l_upper_arm
    [28] l_leg_ft_sensor (dofs: 0) : l_hip_2<-->l_hip_3
    [29] l_foot_ft_sensor (dofs: 0) : l_ankle_2<-->l_foot
    [30] r_leg_ft_sensor (dofs: 0) : r_hip_2<-->r_hip_3
    [31] r_foot_ft_sensor (dofs: 0) : r_ankle_2<-->r_foot
```
</details>

The joints order differs between the reduced and full models. Therefore, [here](https://github.com/robotology/idyntree/blob/aab291d422ef3c314324722149e6e697827f83ce/src/high-level/tests/KinDynComputationsUnitTest.cpp#L628
) from the reduced model it finds the joint `l_arm_ft_sensor` and [here](https://github.com/robotology/idyntree/blob/aab291d422ef3c314324722149e6e697827f83ce/src/high-level/tests/KinDynComputationsUnitTest.cpp#L630) it saves the index `27` (see the printed full model). When using the saved indices here

https://github.com/robotology/idyntree/blob/aab291d422ef3c314324722149e6e697827f83ce/src/high-level/tests/KinDynComputationsUnitTest.cpp#L636-L640

it accesses the elements by index but it saved the index `27` associated with the `l_arm_ft_sensor` which is bigger than the vector size.

Maybe, the joints should be added to the submodel in a different way? Or there is a way to consider only joints of a model that are not fixed joints?

traversaro commented:
> Maybe, the joints should be added to the submodel in a different way? Or there is a way to consider only joints of a model that are not fixed joints?

In general, the dofoffset is different from the joint id. 

I think the correct way to implement this loop:
~~~cxx
 for(size_t jntIdx = 0; jntIdx < reducedModel.getNrOfDOFs(); jntIdx++) 
 { 
     qReducedModel(jntIdx) = qjFullModel(idxJntReducedModelInFullModel(jntIdx)); 
     dqReducedModel(jntIdx) = dqjFullModel(idxJntReducedModelInFullModel(jntIdx)); 
 } 
~~~
is, for positions:
~~~cxx
 for(JointIndex jntIdx = 0; jntIdx < reducedModel.getNrOfJoints(); jntIdx++) 
 {
     size_t posCoordsOffsetReducedModel = reducedModel.getJoint(jntIdx)->getPosCoordsOffset();
     JointIndex jntIdxFullModel = idxJntReducedModelInFullModel(jntIdx);
     size_t posCoordsOffsetFullModel = fullModel.getJoint(jntIdxFullModel)->getPosCoordsOffset();
     for(size_t localPosCoords = 0; localPosCoords < reducedModel.getJoint(jntIdx)->getNrOfPosCoords(); localPosCoords ++) 
     {
         qReducedModel(posCoordsOffsetReducedModel+localPosCoords) = qjFullModel(posCoordsOffsetFullModel+localPosCoords); 
     }
 } 
~~~
while for velocity (similar will apply accelerations/forces and similar quantities):
~~~cxx
 for(JointIndex jntIdx = 0; jntIdx < reducedModel.getNrOfJoints(); jntIdx++) 
 {
     size_t dofsOffsetReducedModel = reducedModel.getJoint(jntIdx)->getDOFsOffset();
     JointIndex jntIdxFullModel = idxJntReducedModelInFullModel(jntIdx);
     size_t dofsOffsetFullModel = fullModel.getJoint(jntIdxFullModel)->getDOFsOffset();
     for(size_t localDofs = 0; localDofs < reducedModel.getJoint(jntIdx)->getNrOfDOFs(); localDofs++) 
     {
         dqReducedModel(dofsOffsetReducedModel+localDofs ) = dqjFullModel(dofsOffsetFullModel+localDofs); 
     }
 } 
~~~

The idea here to have this different ways of is that in the future we could have joints that use for representing positions a different number of variables w.r.t. to velocity (think of a spherical joint in which the positions are quaternions (4D) and velocity are angular velocity (3D).

isorrentino commented:
In addition to your code, I added the check on the DoF of the joint. In this way, I can discard fixed joints. If I do not have this check also positions and velocities of fixed joints are added to the vectors and `setRobotState` gives an error.

I also added tests with all the available models, just to be sure that the test does not fail also in other conditions.

isorrentino commented:
> Thanks! Can I merge with squash?

Sure!

traversaro commented:
This different between 

> In addition to your code, I added the check on the DoF of the joint. In this way, I can discard fixed joints. If I do not have this check also positions and velocities of fixed joints are added to the vectors and `setRobotState` gives an error.

If I am not wrong, this should not be necessary as if there is a 0-dof joint no value is copied as the for loop `for(size_t localDofs = 0; localDofs < reducedModel.getJoint(jntIdx)->getNrOfDOFs(); localDofs++)` exists immediatly.

isorrentino commented:
> > In addition to your code, I added the check on the DoF of the joint. In this way, I can discard fixed joints. If I do not have this check also positions and velocities of fixed joints are added to the vectors and `setRobotState` gives an error.
> 
> If I am not wrong, this should not be necessary as if there is a 0-dof joint no value is copied as the for loop `for(size_t localDofs = 0; localDofs < reducedModel.getJoint(jntIdx)->getNrOfDOFs(); localDofs++)` exists immediatly.

Yes, you are right. I fixed it.

traversaro commented:
Ok, is this ready for merge?

isorrentino commented:
> Ok, is this ready for merge?

Yes, you can merge it.
